### PR TITLE
Fix copying wrapped lines by implementing better scrolling

### LINKED ISSUE #5039: Clear-Host fails to clear the first line
# Environment

```none
Windows build number: 10.0.19041.0
Windows Terminal version: 0.10.761.0

Any other software?
PowerShell Core 7.0.0
NodeJS 12.16.1
coronavirus-tracker-cli 0.6.0 (`npm install -g coronavirus-tracker-cli@0.6.0`)
```

# Steps to reproduce

1. Open PowerShell Core
2. Run `corona` to get output
3. Run `cls` to clear the screen

# Expected behavior

Screen should be fully cleared.

# Actual behavior

The first line retains some characters. This is probably related to the color formatting being used by the CLI. The characters retained are also not from the line that was originally the topmost line.

Output before clearing: 

Output after clearing:

Note: A second `cls` finishes clearing the screen.

### LINKED ISSUE #5113: Terminal breaks copied text at window line wraps, even when the text has no new-line characters
I recall this was working in older versions of Terminal and in Windows' copy of the console window, but it seems to have broken more recently.

### Environment

**Windows build number:** 18363.720 (Windows 10 ver. 1909 with update [KB4551762](https://support.microsoft.com/en-us/help/4551762))
**Windows Terminal version:** 0.10.781.0

### Steps to reproduce

1. Open this file in an editor: `C:\Windows\System32\WindowsCodecsRaw.txt`
1. Notice that the last line of the file "THE SOFTWARE IS PROVIDED".... is over 500 characters long.
1. Open Terminal to the Command Prompt or PowerShell, then run:
  `type C:\Windows\System32\WindowsCodecsRaw.txt`
1. Drag to select text from multiple in-window lines of the last line of the file, then right-click to copy.
1. Paste the clipping into a new file in Notepad.

### Expected and actual behavior

Expected: the text is all on one line.
Actual: the text is split into multiple lines.

### LINKED ISSUE #5161: Scrolling down in VIM scrolls up duplicated vim status lines
# Environment

```none
Windows build number: Microsoft Windows [Version 10.0.18363.720]   
Windows Terminal version (if applicable): 0.10.781.0

vim for Windows (official git build): VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Jan 16 2020 08:26:10)
```

# Steps to reproduce

- Start vim: `vim` (the one that comes with `git version 2.26.0.windows.1`)
- Press `i`
- Press `ENTER` until it starts scrolling the screen
- (to exit vim): `ESC` and then type `:q!` and then `ENTER` 😄

# Expected behavior

The status line stays where it's supposed to be, at the bottom of the screen.

# Actual behavior

Every time a the screen is scrolled down, the last line (white status line) is scrolled up, and a new white status line is printed at the bottom. This results in two white lines, then 3, etc...

This does not reproduce if I use the old cmd.exe (started by eg. Explorer, not wt.exe)

### LINKED ISSUE #5180: Neovim Cursor Flickers from Block to Bar running in MS Terminal under Tmux
# Environment

```none
Windows build number: 10.0.19041.172
Windows Terminal version (if applicable):  0.10.781.0

Other software:
Debian WSL2 running tmux
```

# Steps to reproduce

1. Start tmux
2. Start neovim (nvim -u NORC)
3. Enter insert mode
4. Watch cursor flicker between block and thin line  

# Expected behavior
Cursor should remain a line and not flicker. (this is the behavior in WSLtty)

# Actual behavior
Cursor flickers between block and line.

### PR DESCRIPTION
**Pull request by zadjii-msft**
Now that the Terminal is doing a better job of actually marking which
lines were and were not wrapped, we're not always copying lines as
"wrapped" when they should be. We're more correctly marking lines as not
wrapped, when previously we'd leave them marked wrapped.

The real problem is here in the `ScrollFrame` method - we'd manually
newline the cursor to make the terminal's viewport shift down to a new
line. If we had to scroll the viewport for a _wrapped_ line, this would
cause the Terminal to mark that line as broken, because conpty would
emit an extra `\n` that didn't actually exist.

This more correctly implements `ScrollFrame`. Now, well move where we
"thought" the cursor was, so when we get to the next `PaintBufferLine`,
if the cursor needs to newline for the next line, it'll newline, but if
we're in the middle of a wrapped line, we'll just keep printing the
wrapped line.

A couple follow up bugs were found to be caused by the same bad logic.
See #5039 and #5161 for more details on the investigations there.

## References

* #4741 RwR, which probably made this worse
* #5122, which I branched off of 
* #1245, #357 - a pair of other conpty wrapped lines bugs
* #5228 - A followup issue for this PR

## PR Checklist
* [x] Closes #5113
* [x] Closes #5180 (by fixing DECRST 25)
* [x] Closes #5039
* [x] Closes #5161 (by ensuring we only `removeSpaces` on the actual
  bottom line)
* [x] I work here
* [x] Tests added/passed
* [n/a] Requires documentation to be updated

## Validation Steps Performed

* Checked the cases from #1245, #357 to validate that they still work
* Added more and more tests for these scenarios, and then I added MORE
  tests
* The entire team played with this in selfhost builds

### CODE REVIEW

File: src/renderer/vt/XtermEngine.cpp [hunk 1]
```diff
@@ -105,7 +105,7 @@ XtermEngine::XtermEngine(_In_ wil::unique_hfile hPipe,
         // by prepending a cursor off.
         if (_lastCursorIsVisible)
         {
-            _buffer.insert(0, "\x1b[25l");
+            _buffer.insert(0, "\x1b[?25l");
```
Comment (DHowett-MSFT) [inline src/renderer/vt/XtermEngine.cpp]: .. lol, how long has this been like this, and does it map to any bugs in our repo?

Comment (DHowett-MSFT) [reply src/renderer/vt/XtermEngine.cpp]: and is there a constant val we can use somewhere

File: src/renderer/vt/tracing.cpp [hunk 2]
```diff
@@ -148,7 +148,8 @@ void RenderTracing::TraceStartPaint(const bool quickReturn,
                                     const til::bitmap invalidMap,
                                     const til::rectangle lastViewport,
                                     const til::point scrollDelt,
-                                    const bool cursorMoved) const
+                                    const bool cursorMoved,
+                                    std::optional<short> wrappedRow) const
```
Comment (DHowett-MSFT) [inline src/renderer/vt/tracing.cpp]: take this bb by const ref, even if only to save the copy in the case where tracing is turned off

File: src/renderer/vt/XtermEngine.cpp [hunk 3]
```diff
@@ -359,38 +375,28 @@ try
...
+    // newline, we'll cause any wrapped lines to get broken.
+    //
+    // Instead, we'll just update our internal tracker of where the buffer
+    // contents are. On this frame, we'll then still move the cursor correctly
+    // relative to the new frame contents. To do this, we'll shift our
+    // coordinates we're tracking, like the row that we wrapped on and the
+    // position we think we left the cursor.
+    //
+    // See GH#5113
+    _lastText.Y += dy;
```
Comment (DHowett-MSFT) [inline src/renderer/vt/XtermEngine.cpp]: with the newlining gone, how do we make sure that the pty host knows how many lines got pushed?

File: src/renderer/vt/XtermEngine.cpp [hunk 4]
```diff
@@ -359,38 +375,28 @@ try
     const short dy = _scrollDelta.y<SHORT>();
     const short absDy = static_cast<short>(abs(dy));
```
Comment (DHowett-MSFT) [inline src/renderer/vt/XtermEngine.cpp]: it looks like absDy is unused now

File: src/renderer/vt/XtermEngine.cpp (hunk 3 above)
Comment (DHowett-MSFT) [reply src/renderer/vt/XtermEngine.cpp]: explicit in the comment that was removed, too, is the pushing-into-scrollback behavior

File: src/renderer/vt/XtermEngine.cpp (hunk 1 above)
Comment (zadjii-msft) [reply src/renderer/vt/XtermEngine.cpp]: It was mentioned in #3093, and I had a branch that fixed this before ([dev/migrie/b/be-better-at-hiding](https://github.com/microsoft/terminal/tree/dev/migrie/b/be-better-at-hiding)), but I have no memory of why that was never fixed

File: src/renderer/vt/XtermEngine.cpp (hunk 3 above)
Comment (zadjii-msft) [reply src/renderer/vt/XtermEngine.cpp]: awit could you clarify? maybe it's just 5 o'clock calling, but I can't make sense of this comment 🙃

Comment (DHowett-MSFT) [reply src/renderer/vt/XtermEngine.cpp]: Oh. This function used to print a set of newlines commensurate with the scroll distance, which would make the connected terminal push the scrolled-off lines into _its_ scrollback. The newlines are now dead, and I can't determine whether that broke anything.

**Comment by DHowett** [comment]
Testing a selfhost build with this in it, I am seeing terminal not responding to ?25h. I don't know if this is caused by this PR :) I just haven't done a bisect and wanted to make sure I didn't lose it

File: src/renderer/vt/XtermEngine.cpp (hunk 3 above)
Comment (zadjii-msft) [reply src/renderer/vt/XtermEngine.cpp:390]: They seemingly _didn't_ break anything.

Conpty always emits a frame immediately before every new line, due to TriggerCircling. Basically we'll finish printing the frame before the newline, then we'll additionally paint another frame with `scrollDelta={0, -1}`, and in that second frame, since the cursor is now one line higher, we'll newline to move the cursor down, into the new line in the terminal.

Since we're emitting a frame _every_ newline, I'm not sure this function can actually get called with a delta != 1

**Comment by zadjii-msft** [comment]
> 
> 
> Testing a selfhost build with this in it, I am seeing terminal not responding to ?25h. I don't know if this is caused by this PR :) I just haven't done a bisect and wanted to make sure I didn't lose it

Can I get an exact repro to investigate? If what you're seeing is in 0.11.912.0, I'm not seeing that with a simple 
```sh
printf '\e[?25l'
printf '\e[?25h'
```

in bash

**Comment by DHowett-MSFT** [comment]
**OKAY**, here's your repro.

```
^[[?25l^[[H^[[115CXXXXX^[[4;9H^[[?25h
```

* **terminal size 120x30** (required)
* cursor off
* home the cursor
* write up against the right edge of the screen (X=115, W=120, string="XXXXX")
* _move the cursor again_
* cursor back on

If this arrives in one packet, the cursor will remain hidden.

**Comment by DHowett-MSFT** [comment]
What comes out of the debug tap:

```
␛[?25l␛[1;76HXXXXX␛[?25h
```

it seems like the cursor becomes visible, in delayed-EOL position off the edge. conpty doesn't render out the next cursor movement because no text comes with it. Here it says 76 because my terminal was 80, not 120.

File: src/renderer/vt/XtermEngine.cpp [hunk 5]
```diff
@@ -223,7 +223,21 @@ XtermEngine::XtermEngine(_In_ wil::unique_hfile hPipe,
...
+    // Only skip this paint when we think the cursor is in the cell
+    // immediately off the edge of the terminal, and the actual cursor is in
+    // the last cell of the row. We're in a deferred wrap, but the host
+    // thinks the cursor is actually in-frame.
+    // See ConptyRoundtripTests::DontWrapMoveCursorInSingleFrame
+    const bool cursorIsInDeferredWrap = (nextCursorPosition.X == _lastText.X - 1) && (nextCursorPosition.Y == _lastText.Y);
+    // If all three of these contitions are true, then:
+    //   * cursorIsInDeferredWrap: The cursor is in a position where the line
+    //     filled the last cell of the row, but the host tried to paint it in
+    //     the past cell anyways
```
Comment (DHowett-MSFT) [inline src/renderer/vt/XtermEngine.cpp:236]: `last` cell?

Comment (zadjii-msft) [reply src/renderer/vt/XtermEngine.cpp:236]: frick

**Comment by DHowett-MSFT** [comment]
Alright, found an issue with zsh menu completion

When the screen is scrolled, and the cursor is moved back up using relative positioning, it ends up off-by-one.

### input; newlines are LF, ^M is CR

do this at the bottom of a buffer

```
^[[JFIRST LINE^[[K^M
SECOND LINE^[[K
THIRD LINE^[[2A^M^[[23C<OVERSTRIKE>^[[K^M
^[[1B44444444444^M5555555555^M^[[A^[[A
```

### expected (conhost)

image

### current (terminal)

image

**Comment by DHowett-MSFT** [comment]
debug tap emits:

```
FIRST␣LINE␛[K␍␍␊SECOND␣LINE␍␛[?25l␛[28;24H<OVERSTRIKE>␛[30;1H55555555554␛[K␛[29;1H␛[?25h
```

**Comment by zadjii-msft** [comment]
@DHowett-MSFT thanks for more test cases :) I'll add the above one and the one from #5039 (if I can figure out how to repro that one in a test 😕)

File: src/renderer/vt/XtermEngine.cpp [hunk 6]
```diff
@@ -356,41 +384,75 @@ try
...
-        {
-            std::string seq = std::string(absDy, '\n');
-            hr = _Write(seq);
-            // Mark that the bottom line is new, so we won't spend time with an
-            // ECH on it.
-            _newBottomLine = true;
-        }
-        // We don't need to _MoveCursor the cursor again, because it's still
-        //      at the bottom of the viewport.
+        // Save the old wrap state here. We're gonig to clear it so that
```
Comment (jsoref) [inline src/renderer/vt/XtermEngine.cpp:392]:
```suggestion
        // Save the old wrap state here. We're going to clear it so that
```

**Comment by jsoref** [comment]
@zadjii-msft , personally, I'd suggest adding patterns:
```
TestUtils::VerifyExpectedString\(tb, L"[^"]+"
hostSm.ProcessString(L"[^"]+"
```

File: src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp [hunk 7]
```diff
@@ -1302,3 +1529,280 @@ void ConptyRoundtripTests::ScrollWithMargins()
...
+        TEST_METHOD_PROPERTY(L"Data:paintAfterDECALN", L"{false, true}")
+        TEST_METHOD_PROPERTY(L"Data:changeAttributes", L"{false, true}")
+        TEST_METHOD_PROPERTY(L"Data:useLongSpaces", L"{false, true}")
+        TEST_METHOD_PROPERTY(L"Data:printTextAfterSpaces", L"{false, true}")
+    END_TEST_METHOD_PROPERTIES();
+    constexpr int PaintEveryNewline = 0;
+    constexpr int PaintAfterAllNewlines = 1;
+    constexpr int DontPaintAfterNewlines = 2;
+
+    INIT_TEST_PROPERTY(int, paintEachNewline, L"Ene of manually PaintFrame after each newline is emitted, once at the end of all newlines, or not at all");
```
Comment (jsoref) [inline src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp:1598]: What is `Ene`?

Comment (zadjii-msft) [reply src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp:1598]: that's a typo 😆

File: .github/actions/spell-check/patterns/patterns.txt [hunk 8]
```diff
@@ -5,3 +5,5 @@ https://(?:(?:www\.|)youtube\.com|youtu.be)/[-a-zA-Z0-9?&=]*
 Scro\&ll
 # selectionInput.cpp
  :\\windows\\syste\b
+TestUtils::VerifyExpectedString\(tb, L"[^"]+"
+hostSm.ProcessString(L"[^"]+"
```
Comment (jsoref) [inline .github/actions/spell-check/patterns/patterns.txt:9]: Oops, sorry about that:
```suggestion
hostSm\.ProcessString\(L"[^"]+"
```

File: src/renderer/vt/XtermEngine.cpp [hunk 9]
```diff
@@ -318,10 +347,7 @@ XtermEngine::XtermEngine(_In_ wil::unique_hfile hPipe,
             _lastText = coord;
         }
     }
-    if (_lastText.Y != _lastViewport.ToOrigin().BottomInclusive())
```
Comment (DHowett-MSFT) [inline src/renderer/vt/XtermEngine.cpp:321]: does this check go away because we're checking `_newBottomLine && position is in bottom line` now?

File: src/renderer/vt/XtermEngine.cpp [hunk 10]
```diff
@@ -356,41 +384,92 @@ try
...
+        // when there's more invalid than just the bottom line. If only the
+        // bottom line is invalid, then the next thing the Renderer is going to
+        // tell us to do is print the new line at the bottom of the viewport,
+        // and _MoveCursor will automatically give us the newline we want.
+        // When that's implemented, we'll probably want to make sure to add a
+        //   _lastText.Y += dy;
+        // statement here.
+
+        // Move the cursor to the bottom of the current viewport
+        const short bottom = _lastViewport.BottomInclusive();
```
Comment (DHowett-MSFT) [inline src/renderer/vt/XtermEngine.cpp:410]: the loss of `ToOrigin` worries me.

File: src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp [hunk 11]
```diff
@@ -1279,8 +1507,9 @@ void ConptyRoundtripTests::ScrollWithMargins()
         expectedOutput.push_back("\r\n");
     }
     {
-        const std::string expectedString(initialTermView.Width(), '*');
-        expectedOutput.push_back(expectedString);
+        const std::string expectedString(initialTermView.Width() - 1, '*');
+        // There will be one extra blank space at the end of the line, because the
```
Comment (DHowett-MSFT) [inline src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp:1511]: because the what!

File: src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp [hunk 12]
```diff
@@ -1302,3 +1531,380 @@ void ConptyRoundtripTests::ScrollWithMargins()
...
+    Log::Comment(L"========== Checking the terminal buffer state ==========");
+
+    verifyBuffer(termTb, term->_mutableViewport.ToInclusive());
+}
+
+void ConptyRoundtripTests::MarginsWithStatusLine()
+{
+    // See https://github.com/microsoft/terminal/issues/5161
+    //
+    // This test reproduces a case from the MSYS/cygwin vim. From what I can
```
Comment (DHowett-MSFT) [inline src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp:1818]: call out explicitly, cygwin runtime <3.1 (3.1 uses VT now (!))

**Comment by jsoref** [comment]
@zadjii-msft: the easiest way to run it right now is to just create a new branch (not attached to a PR) and push to it. The bot will run and report on the commit w/o spamming any PRs (since there isn't one).

You can run it in a docker instance locally. You basically clone https://github.com/check-spelling/check-spelling/tree/master (you can pick a version, as a general rule, `master` will point to the latest release, which in general should be roughly equivalent to the release microsoft/terminal is running -- I do testing on a branch called `prerelease`), run:

```
docker build . --tag spellchecker:alpha
docker run -it --entrypoint /bin/bash spellchecker:alpha
/usr/bin/git init /home/runner/work/terminal/terminal
cd /home/runner/work/terminal/terminal
/usr/bin/git remote add origin https://github.com/microsoft/terminal
/usr/bin/git config --local gc.auto 0
/usr/bin/git -c protocol.version=2 fetch --no-tags --prune --progress --no-recurse-submodules --depth=5 origin +55f81bc185b6985dcc268f2b1b2dcad9e3753db3:refs/remotes/origin/dev/migrie/b/5113-with-miniksas-fix
# this you'd vary this line you can fetch however your would normally fetch, I'm copying the command from https://github.com/microsoft/terminal/runs/564889539
/usr/bin/git checkout --progress --force -B dev/migrie/b/5113-with-miniksas-fix refs/remotes/origin/dev/migrie/b/5113-with-miniksas-fix
# ibid
export bucket=.github/actions
export project=spell-check
/app/unknown-words.sh
```

I'm open to advice about how best to make this easier for people to do (the `bucket` and `project` fields come from https://github.com/microsoft/terminal/blob/master/.github/workflows/spelling.yml)

One could probably wrap up the above into a different Dockerfile.

File: src/renderer/vt/XtermEngine.cpp (hunk 9 above)
Comment (zadjii-msft) [reply src/renderer/vt/XtermEngine.cpp:321]: Nah I think it was just a tad bit to defensive on the original implementation of the `_newBottomLine` optimization. Now that we're doing

```c++
    if (coord.Y == _lastViewport.BottomInclusive())
    {
        _newBottomLine = false;
    }
```

in paint.cpp, this old check would clear out the `_newBottomLine` state _too_ often. This one (combined with the paint.cpp change) combine to fix #5039

File: .github/actions/spell-check/whitelist/alphabet.txt [hunk 13]
```diff
@@ -1,3 +1,4 @@
+AAAAAAAAAA
```
Comment (miniksa) [inline .github/actions/spell-check/whitelist/alphabet.txt:1]: Man, maybe we need the pattern of 
```
([A-Za-z])\1{3,}
```
To match any letter repeated 4 or more times.

File: src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp [hunk 14]
```diff
@@ -48,6 +49,11 @@ namespace TerminalCoreUnitTests
 };
 using namespace TerminalCoreUnitTests;
 
+// Helper for declaring a variable to store a TEST_METHOD_PROPERTY and get it's value from the test metadata
+#define INIT_TEST_PROPERTY(type, identifer, description) \
+    type identifer;                                      \
+    VERIFY_SUCCEEDED(TestData::TryGetValue(L#identifer, identifer), description);
+
```
Comment (miniksa) [inline src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp:56]: Man this helper should totally bubble up somewhere higher level than this.

File: src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp [hunk 15]
```diff
@@ -174,6 +180,14 @@ class TerminalCoreUnitTests::ConptyRoundtripTests final
 
     TEST_METHOD(TestResizeHeight);
 
+    TEST_METHOD(OutputWrappedLinesAtTopOfBuffer);
+    TEST_METHOD(OutputWrappedLinesAtBottomOfBuffer);
+    TEST_METHOD(ScrollWithChangesInMiddle);
+    TEST_METHOD(DontWrapMoveCursorInSingleFrame);
+    TEST_METHOD(ClearHostTrickeryTest);
+    TEST_METHOD(OverstrikeAtBottomOfBuffer);
+    TEST_METHOD(MarginsWithStatusLine);
```
Comment (miniksa) [inline src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp:189]: OK so I was mildly concerned with the complexity of the trickery going on in the `XtermEngine` prior to looking at this. But if you wrote this many tests around it, I'm now less concerned.

File: src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp [hunk 16]
```diff
@@ -1061,6 +1074,225 @@ void ConptyRoundtripTests::PassthroughHardReset()
...
+    auto verifyBuffer = [](const TextBuffer& tb, const short wrappedRow) {
+        VERIFY_IS_TRUE(tb.GetRowByOffset(wrappedRow).GetCharRow().WasWrapForced());
+        VERIFY_IS_FALSE(tb.GetRowByOffset(wrappedRow + 1).GetCharRow().WasWrapForced());
+
+        auto iter0 = tb.GetCellDataAt({ 0, wrappedRow });
+        TestUtils::VerifySpanOfText(L"A", iter0, 0, TerminalViewWidth);
+        auto iter1 = tb.GetCellDataAt({ 0, wrappedRow + 1 });
+        TestUtils::VerifySpanOfText(L"A", iter1, 0, 20);
+        auto iter2 = tb.GetCellDataAt({ 20, wrappedRow + 1 });
+        TestUtils::VerifySpanOfText(L" ", iter2, 0, TerminalViewWidth - 20);
```
Comment (miniksa) [inline src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp:1180]: You're going to mildly hate me for this request, but...
This is difficult to visualize.

Would you mind creating a large ASCII comment here that shows what the frame looks like before/after your operations? I feel like it might be useful to have it visualized for whomever breaks this in the future to understand what's going on.

Like
```
// H = Hard newline. We don't wrap.
// S = Soft newline. We wrap.

// We set up the initial buffer to be a field of Xs and the cursor is here:
// XXXXXXXXXXX (H)
// XXXXXXXXXXX (H)
// XXXXXXXXXXX (H)
// XXXXXXXXXXX (S)
             ^
             \---- The cursor is sitting on the bottom right one.
...

// Then a scroll frame occurs with a newline. We lost the top line of Xs and it looks like this:
// XXXXXXXXXXX (H)
// XXXXXXXXXXX (H)
// XXXXXXXXXXX (S)
// 
// ^
// \--- The cursor is now here after the \r\n


...

// Then we move the cursor back to the last character of the wrapped row
// XXXXXXXXXXX (H)
// XXXXXXXXXXX (H)
// XXXXXXXXXXX<-- The cursor is now here after the implicit move (S)
// 

...

// Then we put in a bunch of As
// It will fill the spot under the cursor first...
// XXXXXXXXXXX (H)
// XXXXXXXXXXX (H) 
// XXXXXXXXXXA<-- The cursor is still here, but due to the wrap, we won't get a newline when PTY draws (S)
// 
// 
// And then continue on the row below
// XXXXXXXXXXX
// XXXXXXXXXXX
// XXXXXXXXXXA
// AAAAAAAAA
//          ^
//           \-The cursor ends up here.

File: src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp [hunk 17]
```diff
@@ -1061,6 +1074,225 @@ void ConptyRoundtripTests::PassthroughHardReset()
...
+        VERIFY_SUCCEEDED(renderer.PaintFrame());
+    }
+
+    const auto wrappedLineLength = TerminalViewWidth + 20;
+
+    expectedOutput.push_back(std::string(TerminalViewWidth, 'A'));
+    // TODO GH#5228 might break the "newline & repaint the wrapped char" checks here, that's okay.
+    expectedOutput.push_back("\r"); // This \r\n is emitted by ScrollFrame to
+    expectedOutput.push_back("\n"); // add a newline to the bottom of the buffer
+    expectedOutput.push_back("\x1b[31;80H"); // Move the cursor BACK to the wrapped row
```
Comment (miniksa) [inline src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp:1165]: row 31, 80.... this is implicit based on the global state? I'm not a huge fan of that... You might want to go warn heavily where the global state is set up that it cannot be changed from 80x32 because some of the tests presume that in their hardcoded VT.

File: .github/actions/spell-check/whitelist/alphabet.txt (hunk 13 above)
Comment (jsoref) [reply .github/actions/spell-check/whitelist/alphabet.txt:1]: If that works, go for it!
I don't use regexps that fancy.

Comment (DHowett-MSFT) [reply .github/actions/spell-check/whitelist/alphabet.txt:1]: Oh god, regexes with backreferences. I get that we want to do it because they have to be the same character, but man, that's scary.

**Comment by DHowett-MSFT** [comment]
You're going to hate me, but when a newly-printed line (as in #5113) wraps _on a space character_, we don't properly copy-as-wrapped on the terminal-side

**Comment by DHowett-MSFT** [comment]
From PowerShell,

```powershell
write-host "$("A" * ($host.ui.rawui.windowsize.width-2))   This should have been one line"
```

File: src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp (hunk 16 above)
Comment (zadjii-msft) [reply src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp:1180]: No problemo. I did a better job in the later tests, but yea these need love too

**Comment by zadjii-msft** [comment]
@DHowett-MSFT I don't hate you, I love test cases. The more tests, the better I feel about this

File: .github/actions/spell-check/patterns/patterns.txt [hunk 18]
```diff
@@ -7,3 +7,4 @@ Scro\&ll
  :\\windows\\syste\b
 TestUtils::VerifyExpectedString\(tb, L"[^"]+"
 hostSm\.ProcessString\(L"[^"]+"
+([A-Za-z])\1{3,}
```
Comment (jsoref) [inline .github/actions/spell-check/patterns/patterns.txt:10]: This seems to work reasonably well in this repository.

For powerdns/pdns i'm going to add `\b` around this line.

Otherwise, a `aaarec` which is a typo would be accepted along w/ `aaaarec` which is not a typo.

Comment (miniksa) [reply .github/actions/spell-check/patterns/patterns.txt:10]: Oh, that's actually a good call. This should probably be book-ended in `\b`s so it only hits full words and breaks on boundaries.

File: src/renderer/vt/paint.cpp [hunk 19]
```diff
@@ -558,7 +555,7 @@ using namespace Microsoft::Console::Types;
         {
             _deferredCursorPos = { _lastText.X + sNumSpaces, _lastText.Y };
         }
-        else if (numSpaces > 0)
+        else if (removeSpaces && numSpaces > 0)
```
Comment (DHowett-MSFT) [inline src/renderer/vt/paint.cpp:558]: If we're removing spaces, and there were spaces, print them? I am confus

Comment (zadjii-msft) [reply src/renderer/vt/paint.cpp:558]: Man so am I, I have no idea why that fixed the test and didn't break any of the other ones. It was literally the first thing I guessed, looks like I didn't actually add a comment

File: src/renderer/vt/paint.cpp [hunk 20]
```diff
@@ -538,15 +547,15 @@ using namespace Microsoft::Console::Types;
             RETURN_IF_FAILED(_EraseLine());
         }
     }
-    else if (_newBottomLine)
+    else if (_newBottomLine && coord.Y == _lastViewport.BottomInclusive())
```
Comment (DHowett-MSFT) [inline src/renderer/vt/paint.cpp:550]: oh it was the same issue as before! should we pull this out into a local, `dirtyingNewBottomLine`? That'll make sure we don't use _newBottomLine as much in the body here..

Comment (zadjii-msft) [reply src/renderer/vt/paint.cpp:550]: Eh probably

**Comment by DHowett-MSFT** [comment]
@msftbot merge this in 5 minutes
